src/serializer/serializer.js

```diff
@@ -132,11 +133,20 @@ export class Serializer {
 
     // Deep traversal of the heap to identify the necessary scope of residual functions
     if (timingStats !== undefined) timingStats.deepTraversalTime = Date.now();
+    let preludeGenerator = this.realm.preludeGenerator;
+    invariant(preludeGenerator !== undefined);
+    let referentializer = new Referentializer(
```

Comment:
So there used to be a separate Referentializer for each ResidualHeapSerializer instance (they will be two when we run with "inlineExpressions"). Could that cause any complications with state that gets shared? Maybe it's okay, but it's hard to tell. Making as many details of the Referentializer private (prefixing with "_") would help to tell.

The first issue I see is around statistics. Looking at the code, I just realized that many months ago the code got refactored to share a single SerializerStatistics instance --- that doesn't look right.

Effectively doing referentialization only once instead of twice (with "inlineExpressions") would be nice, as it will make things faster.

Reply:
It does cause some complications which are taken care of in `cleanInstance` in the Referentializer invoked at the end of serialization.